### DISCUSSION TITLE
Lesson 7 MockV3Aggregator Question

### DISCUSSION THREAD

**Discussion by SCPassion** [Cyfrin/foundry-full-course-cu#626] [Q&A] [answered]
I would like to understand more regarding what the MockV3Aggregator is doing behind the scene. ( as we just copied the code from the Foundry fund me project on github)

When we test with the PriceFeedVersionIsAccurate() function in the test script.
```
    function testPriceFeedVersionIsAccurate() public {
        // This is a unit test
        uint256 version = fundMe.getVersion();
        assertEq(version, 4);
    }
```
As of my understanding, it is actually call the following function in FundMe.sol
```
    function getVersion() public view returns (uint256) {
        return s_priceFeed.version();
    }
```
Whereas, s_priceFeed is defined as:
```
AggregatorV3Interface private s_priceFeed;

```
Which direct to here
```
import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol"; // import from github
```

How does the HelperConfig.s.sol resolve the issue? Is s_priceFeed actually MockV3Aggregator instead of AggregatorV3Interface when working with anvil?

Thanks for reading my post. I am looking forward to hearing some comments.

**Comment by Tanmay53** [comment, marked as answer]
The **s_price_feed** storage variable gets defined in the constructor function at line 48 in [FundMe.sol](https://github.com/Cyfrin/foundry-fund-me-f23/blob/main/src/FundMe.sol)

```solidity
    constructor(address priceFeed) {
        s_priceFeed = AggregatorV3Interface(priceFeed);
        i_owner = msg.sender;
    }
```

The constructor accepts the **priceFeed** address, which in case of **Anvil**(deploying locally) is the address to the **MockV3Aggregator**.

Now, the HelperConfig.s.sol comes into play which deploys as well as provides the address for **MockV3Aggregator** when we deploy the ***FundMe*** contract using the DeployScript.s.sol.

**Reply by SCPassion** [reply]
Thanks for answering. But, how does anvil know what to do with AggregatorV3Interface? This interface is from chainlink.

**Reply by SCPassion** [reply]
Aha. I get what you mean. We are missing the address of Price feed for the local anvil blockchain. With the help of DeployFundMe.s.sol, we create a address for each blockchain and pass the corresponding address to create the FundMe smart contract. The AggregatorV3 is already installed before with smartcontractkit/chainlink-brownie-contracts.

Thanks for your explanation.

**Reply by Tanmay53** [reply]
Great to help.

You can also mark the first comment as the answer for this question.

**Comment by theirrationalone** [comment]
Yes, `s_priceFeed` is an instance of `MockV3Aggregator`. This means that `s_priceFeed` is just an object of `AggregatorV3`. When we work or test locally, or on Anvil, we need something or someone capable of imitating an `AggregatorV3`. Therefore, `MockV3Aggregator` is implemented in a way that allows it to adopt all properties and functionalities of `AggregatorV3`. Essentially, `MockV3Aggregator` acts as a stand-in for `AggregatorV3`, or more accurately, I would refer to it as the "PriceFeed Aggregator Version 3".

In the context of being on a local chain, let's say on Anvil, the code in the constructor helps to switch to the network associated with the provided `chain.id`.

So, if we're on Anvil, then `getOrCreateAnvilEthConfig` is invoked.

`HelperConfig.s.sol`:...

```solidity
constructor() {
    if (block.chainid == 11155111) {
        activeNetworkConfig = getSepoliaEthConfig();
    } else {
        activeNetworkConfig = getOrCreateAnvilEthConfig(); // called when Anvil is active or when we're testing locally through Anvil.
    }
}

```

Now, let's delve into getOrCreateAnvilEthConfig:

If we already have an instance of PriceFeed, we simply return the past networkConfigs.

Next, vm.broadcast is used to announce that MockV3Aggregator is active in a temporary Anvil (spawned by Foundry) or Anvil (spawned by you), making it accessible. After that,

We create an instance of MockV3Aggregator that mimics an AggregatorV3. Then, we store the address of this newly created MockV3Aggregator instance also called `PriceFeed` into `activeNetworkConfig` storage variable of custom type `NetworkConfig`.

`HelperConfig.s.sol`:...

```solditiy
function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory anvilNetworkConfig) {
    if (activeNetworkConfig.priceFeed != address(0)) {
        return activeNetworkConfig;
    }
    vm.startBroadcast();
    MockV3Aggregator mockPriceFeed = new MockV3Aggregator(
        DECIMALS,
        INITIAL_PRICE
    );
    vm.stopBroadcast();

    anvilNetworkConfig = NetworkConfig({priceFeed: address(mockPriceFeed)});
}
```

Subsequently, when deploying our contract using the DeployFundMe script, we also invoke the Helperconfig function to obtain network configurations.

We retrieve the priceFeed from an instance of HelperConfig by invoking the activeNetworkConfig function, which supplies the priceFeed address in return. And guess what? This PriceFeed is precisely what we require in our contract.

`FuneMe.sol`:.... **Also read Comments**

```solidity
constructor(address priceFeed) {
        s_priceFeed = AggregatorV3Interface(priceFeed);    // <- Here it Doesn't create an instance of AggregatorV3Interface, Instead of this, It gives gives an already created instance of PriceFeed that facilitates to retrieve / interact with this contract using PriceConverter.sol Library.
        i_owner = msg.sender;
    }
```

`DeployFundMe.s.sol`:....

```solidity
contract DeployFundMe is Script {
    function run() external returns (FundMe, HelperConfig) {
        HelperConfig helperConfig = new HelperConfig(); // This comes with our mocks!
        address priceFeed = helperConfig.activeNetworkConfig();

        vm.startBroadcast();    // And likewise here also broadcast make the instance of FundMe, available for usage.
        FundMe fundMe = new FundMe(priceFeed);    // <- Here we Put the priceFeed of type address.
        vm.stopBroadcast();
        return (fundMe, helperConfig);
    }
}
```

Please correct me if i went somewhere wrong ❌

Hope It Helps 😉 

Thank You 🙏
